src/EntityId/PrefixingEntityIdParser.php

```diff
@@ -0,0 +1,56 @@
+<?php
+
+namespace Wikibase\DataModel\Services\EntityId;
+
+use Wikimedia\Assert\Assert;
+use Wikibase\DataModel\Entity\EntityId;
+use Wikibase\DataModel\Entity\EntityIdParser;
+
+/**
+ * EntityIdParser that adds a fixed prefix and parses resulting string as an EntityId.
+ * This can be used to prefix IDs of entities coming from a foreign repository
+ * with the repository name to avoid clashes with IDs of local entities.
+ *
+ * @since 3.7
+ *
+ * @license GPL-2.0+
+ */
+class PrefixingEntityIdParser implements EntityIdParser {
+
+	/**
+	 * @var string
+	 */
+	private $prefix;
+
+	/**
+	 * @var EntityIdParser
+	 */
+	private $idParser;
+
+	/**
+	 * @param string $prefix Prefix to be added
+	 * @param EntityIdParser $idParser
+	 */
+	public function __construct( $prefix, EntityIdParser $idParser ) {
+		Assert::parameterType( 'string', $prefix, '$prefix' );
+		$this->prefix = $prefix;
+		$this->idParser = $idParser;
+	}
+
+	/**
+	 * Adds the prefix to the given string and parses the resulting string as an EntityId.
+	 * If the string is already prefixed, the prefix is not duplicated (this check is case sensitive).
+	 *
+	 * @param string $idSerialization
+	 * @return EntityId
+	 */
+	public function parse( $idSerialization ) {
+		// Do not add prefix if $idSerialization already starts with it
+		if ( strncmp( $idSerialization, $this->prefix, strlen( $this->prefix ) ) !== 0 ) {
```

Comment:
Oh, I completely missed this, and I think this is the cause for Thiemo's criticism: this is wrong. The prefix must ALWAYS by applied. If the repo we call "foo" sends us an Id with the prefix "foo" we can *not* assume that that's the same foo. We must add the prefix anyway, resulting in "foo:foo:Q123", meaning "Q123 in the repo that the repo we call foo calls foo". Confusing, but consistent :)